### PR TITLE
Replace uses of evaluate* and get* with evaluate() and value(). 

### LINKED ISSUE #643: Cleanup legacy methods in `Val`
Current, `Val` has the following methods:
```C++
  // If this Val is an integer with a direct constant value associated with it,
  // will return the value of that constant integer. If this integer has
  // defining expressions it will return a std::nullopt. Those values should be
  // infered using evaluateInt.
  std::optional<int64_t> getInt() const;

  // If this Val is a double with a direct constant value associated with it,
  // will return the value of that constant double. If this double has
  // defining expressions it will return a std::nullopt. Those values should be
  // infered using evaluateDouble.
  std::optional<double> getDouble() const;

  // If this Val is a bool with a direct constant value associated with it,
  // will return the value of that constant bool. If this bool has defining
  // expressions it will return a std::nullopt. Those values should be infered
  // using evaluateBool.
  std::optional<bool> getBool() const;

  // If this Val is a constant integer, and its history is comprised only of
  // constant values, will return the value of that constant integer. Cannot
  // make constant as expression evaluator takes non-constant Vals.
  int64_t evaluateInt();

  // If this Val is a constant double, and its history is comprised only of
  // constant values, will return the value of that constant double. Cannot
  // make constant as expression evaluator takes non-constant Vals.
  double evaluateDouble();

  // If this Val is a constant bool, and its history is comprised only of
  // constant values, will return the value of that constant bool. Cannot
  // make constant as expression evaluator takes non-constant Vals.
  bool evaluateBool();
```
This was a legacy before we have a `PolymorphicValue`. With `PolymorphicValue`, we only need a single
```C++
const PolymorphicValue& value() const;
```
and a single
```C++
PolymorphicValue evaluate();
```
we should replace all uses of the above 6 legacy methods with the two new methods.

### PR DESCRIPTION
**Pull request by protonu**
This replaces uses of evaluate(Int, Bool, Double) with evaluate which returns a PolymorphicValue.

This also replaces the uses of get(Int, Bool, Double) with value().

Fixes #643

### CODE REVIEW

File: csrc/evaluator_common.cpp [hunk 1]
```diff
@@ -190,14 +190,10 @@ void PrecomputedValues::initializeValueList(
...
-      if (sorted_value_list[i]->isFloatingPointScalar()) {
-        values_[i] = PolymorphicValue(sorted_value_list[i]->evaluateDouble());
-      }
-      if (sorted_value_list[i]->isABool()) {
-        values_[i] = PolymorphicValue(sorted_value_list[i]->evaluateBool());
+      if (sorted_value_list[i]->isIntegralScalar() ||
+          sorted_value_list[i]->isFloatingPointScalar() ||
+          sorted_value_list[i]->isABool()) {
+        values_[i] = PolymorphicValue(sorted_value_list[i]->evaluate());
       }
```
Comment (zasdfgbnm) [inline csrc/evaluator_common.cpp:197]: I think we can just do
```suggestion
      values_[i] = sorted_value_list[i]->evaluate();
```
In the past, `evaluate` only supports these three dtypes, that's why we have that check. But this is no longer the case, so we should just remove these checks.

File: csrc/expr_simplifier.cpp [hunk 2]
```diff
@@ -273,19 +273,10 @@ Val* foldConstants(Val* value) {
     return value;
   }
   if (value->isConstScalar()) {
-    if (value->isIntegralScalar()) {
-      return IrBuilder::create<Val>(
-          value->evaluateInt(), *value->getDataType());
+    if (value->isIntegralScalar() || value->isFloatingPointScalar() ||
```
Comment (zasdfgbnm) [inline csrc/expr_simplifier.cpp:276]: Similar to https://github.com/NVIDIA/Fuser/pull/1104/files#r1364136975

**Comment by wujingyue** [comment]
Thanks for the PR, @protonu ! For better tracking, I modified the title to be more descriptive and linked the issue in the commit message.

File: csrc/evaluator_common.cpp (hunk 1 above)
Comment (protonu) [reply csrc/evaluator_common.cpp]: Thanks! Adressed.

File: csrc/evaluator_common.cpp [hunk 3]
```diff
@@ -190,15 +190,7 @@ void PrecomputedValues::initializeValueList(
...
-      if (sorted_value_list[i]->isIntegralScalar()) {
-        values_[i] = PolymorphicValue(sorted_value_list[i]->evaluateInt());
-      }
-      if (sorted_value_list[i]->isFloatingPointScalar()) {
-        values_[i] = PolymorphicValue(sorted_value_list[i]->evaluateDouble());
-      }
-      if (sorted_value_list[i]->isABool()) {
-        values_[i] = PolymorphicValue(sorted_value_list[i]->evaluateBool());
-      }
+      values_[i] = PolymorphicValue(sorted_value_list[i]->evaluate());
```
Comment (zasdfgbnm) [inline csrc/evaluator_common.cpp:193]: `evaluate` already returns `PolymorphicValue`, so I don't think we need this constructor call.

Comment (protonu) [reply csrc/evaluator_common.cpp]: Thanks! I should have noticed that

File: csrc/parallel_dimension_map.cpp [hunk 4]
```diff
@@ -74,9 +74,10 @@ void ParallelDimensionMap::build(Fusion* fusion) {
 
   // Compute exact_types_
   for (auto [ptype, concrete_id] : all_concrete_ids) {
-    if (simplifyExpr(SimplifyingIrBuilder::eqExpr(
-                         dim_map_.at(ptype), concrete_id->extent()))
-            ->getBool() != true) {
+    auto expr_val = simplifyExpr(SimplifyingIrBuilder::eqExpr(
+                                     dim_map_.at(ptype), concrete_id->extent()))
+                        ->value();
+    if (expr_val.hasValue() && expr_val != true) {
```
Comment (zasdfgbnm) [inline csrc/parallel_dimension_map.cpp:80]: nit:
```suggestion
    if (expr_val.hasValue() && !expr_val.as<bool>()) {
```

File: csrc/parallel_dimension_map.cpp [hunk 5]
```diff
@@ -109,11 +110,14 @@ void ParallelDimensionMap::adjustMappingsForWarpPadding() {
...
-  if (simplifyExpr(SimplifyingIrBuilder::eqExpr(
+  auto expr_val =
+      simplifyExpr(SimplifyingIrBuilder::eqExpr(
                        SimplifyingIrBuilder::modExpr(tidx_dim, warp_size_val),
                        tidx_dim->container()->zeroVal()))
-          ->getBool() == true) {
+          ->value();
+
+  // If already multiple of warp, nothing to do
+  if (expr_val.hasValue() && expr_val == true) {
```
Comment (zasdfgbnm) [inline csrc/parallel_dimension_map.cpp:120]: nit:
```suggestion
  if (expr_val.hasValue() && expr_val.as<bool>()) {
```

File: test/test_expr_simplifier.cpp [hunk 6]
```diff
@@ -30,7 +30,7 @@ namespace {
 // simplifies to true. We don't use x->sameAs(y), because we want to consider
 // a(b+c), ab+ac, (b+c)a, ba+ca as equivalent, but `sameAs` can not do this job.
 bool isEquivalent(Val* x, Val* y) {
-  return simplifyExpr(IrBuilder::eqExpr(x, y))->getBool() == true;
+  return simplifyExpr(IrBuilder::eqExpr(x, y))->value() == true;
```
Comment (zasdfgbnm) [inline test/test_expr_simplifier.cpp:33]: nit:
```suggestion
  return simplifyExpr(IrBuilder::eqExpr(x, y))->value().as<bool>();
```

**Comment by protonu** [comment]
Thanks @zasdfgbnm! I think I addresses your comments. Please take another look

File: csrc/parallel_dimension_map.cpp [hunk 7]
```diff
@@ -109,11 +110,14 @@ void ParallelDimensionMap::adjustMappingsForWarpPadding() {
...
-  if (simplifyExpr(SimplifyingIrBuilder::eqExpr(
+  auto expr_val =
+      simplifyExpr(SimplifyingIrBuilder::eqExpr(
                        SimplifyingIrBuilder::modExpr(tidx_dim, warp_size_val),
                        tidx_dim->container()->zeroVal()))
-          ->getBool() == true) {
+          ->value();
+
+  // If already multiple of warp, nothing to do
+  if (expr_val.hasValue() && expr_val.as<bool>() == true) {
```
Comment (zasdfgbnm) [inline csrc/parallel_dimension_map.cpp:120]: Nit:
```suggestion
  if (expr_val.hasValue() && expr_val.as<bool>()) {
```

File: csrc/ir/base_nodes.cpp [hunk 8]
```diff
@@ -206,115 +206,44 @@ bool Val::isConstInt() const {
...
 
 bool Val::isOneInt() const {
-  auto int_val = getInt();
-  return int_val.has_value() && int_val.value() == 1;
+  return value().hasValue() && value().is<int64_t>() && value() == 1;
 }
 
 bool Val::isTrue() const {
-  return getBool() == true;
+return value().hasValue() && value().as<bool>();
```
Comment (zasdfgbnm) [inline csrc/ir/base_nodes.cpp:242]:
```suggestion
return value().hasValue() && value().is<bool>() && value().as<bool>();
```

File: csrc/ir/base_nodes.cpp [hunk 9]
```diff
@@ -206,115 +206,44 @@ bool Val::isConstInt() const {
...
-    return std::nullopt;
-  }
-  return std::nullopt;
+  return evaluated_val;
 }
 
 bool Val::isZero() const {
-  return getInt() == 0 || getDouble() == 0.0;
+  return (
+      this->value().hasValue() && (this->value() == 0 || this->value() == 0.0));
```
Comment (zasdfgbnm) [inline csrc/ir/base_nodes.cpp:225]:
```suggestion
      this->value().hasValue() && this->value() == 0);
```

File: csrc/ir/base_nodes.h [hunk 10]
```diff
@@ -326,38 +326,11 @@ class Val : public Statement {
...
-  // If this Val is a constant double, and its history is comprised only of
-  // constant values, will return the value of that constant double. Cannot
-  // make constant as expression evaluator takes non-constant Vals.
-  double evaluateDouble();
-
-  // If this Val is a constant bool, and its history is comprised only of
-  // constant values, will return the value of that constant bool. Cannot
-  // make constant as expression evaluator takes non-constant Vals.
-  bool evaluateBool();
+  // If this Val is a constant type (such as int, double, bool ...), and its
```
Comment (zasdfgbnm) [inline csrc/ir/base_nodes.h:329]:
```suggestion
  // If this Val's history
```
dtype does not matter here. We support all dtypes now.

File: csrc/expr_simplifier.cpp [hunk 11]
```diff
@@ -1490,10 +1478,8 @@ bool isPositiveHelper(Val* value, const Context& context) {
 
 bool isNonZero(Val* value, const Context& context) {
   value = foldConstants(value);
-  if (value->getInt().has_value() && *value->getInt() != 0) {
-    return true;
-  }
-  if (value->getDouble().has_value() && *value->getDouble() != 0.0) {
+  if (value->value().hasValue() && value->value().is<int64_t>() &&
+      value->value().as<int64_t>() != 0) {
```
Comment (zasdfgbnm) [inline csrc/expr_simplifier.cpp:1482]: I think this should be
```suggestion
  if (value->value().hasValue() && value->value() != 0) {
```

File: csrc/expr_simplifier.cpp [hunk 12]
```diff
@@ -1557,11 +1541,9 @@ bool lessThan(Val* x, Val* y, const Context& context) {
 bool lessEqual(Val* x, Val* y, const Context& context) {
   x = foldConstants(x);
   y = foldConstants(y);
-  if (x->getInt().has_value() && y->getInt().has_value()) {
-    return *x->getInt() <= *y->getInt();
-  }
-  if (x->getDouble().has_value() && y->getDouble().has_value()) {
-    return *x->getDouble() <= *y->getDouble();
+  if (x->value().hasValue() && x->value().is<int64_t>() &&
+      y->value().hasValue() && y->value().is<int64_t>()) {
```
Comment (zasdfgbnm) [inline csrc/expr_simplifier.cpp:1545]: I think this needs to be
```suggestion
  if (x->value().hasValue() && y->value().hasValue()) {
```

**Comment by protonu** [comment]
I'll start splitting this PR to smaller PRs and remove this PR. The first to replace this will be #1155

**Pull request closed without merging**